File: internal/cli/testdata/golden/nomnom-ts/tracking/src/nomnom/tracking.ts
// Tracking: where an order is up to.
//
// The one unit that is reached both ways. It *subscribes* to courierAssigned,
// because a courier having been assigned is a statement nobody waits on, and it
// is *called* by the storefront, because a customer asking where their food is
// wants an answer now.
//
// Both arrive at the same Lambda and the generated entrypoint tells them apart
// by the shape of the event. Nothing in this file has to know that.

// Injected by cloudcc: runtime clients for this program's declared capabilities.
import * as _cloudccKv from "../../_cloudcc_runtime/kv.js";

import { DynamoDBClient, GetItemCommand, PutItemCommand } from "@aws-sdk/client-dynamodb";

import { courierAssigned, type CourierAssigned } from "./events";

undefined;

const TABLE = "nomnom-tracking" as const;

const timeline = _cloudccKv.connect("trackingEvents") as DynamoDBClient;

export interface OrderStatus {
  order_id: string;
  state: string;
  courier: string | null;
}

/**
 * A subscriber, not a remote function: nobody is waiting on the result.
 *
 * The compiler only requires `async` of functions something calls over the
 * wire, which this is not -- the message arrives on its own. It is `async` here
 * only because writing to DynamoDB is.
 */
async function onCourierAssigned(message: CourierAssigned) {
  const orderId = message.order_id ?? "unknown";
  await timeline.send(
    new PutItemCommand({
      TableName: TABLE,
      Item: {
        id: { S: orderId },
        state: { S: "out-for-delivery" },
        courier: { S: message.courier ?? "unassigned" },
      },
    }),
  );
  return { tracked: orderId };
}

courierAssigned.subscribe(onCourierAssigned);

/** Called by the storefront while a customer is looking at the page. */
export async function orderStatus(orderId: string): Promise<OrderStatus> {
  const out = await timeline.send(
    new GetItemCommand({ TableName: TABLE, Key: { id: { S: orderId } } }),
  );
  if (!out.Item) {
    return { order_id: orderId, state: "preparing", courier: null };
  }
  return {
    order_id: orderId,
    state: out.Item.state!.S!,
    courier: out.Item.courier?.S ?? null,
  };
}

File: internal/cli/testdata/golden/nomnom/tracking/nomnom/tracking.py
"""Tracking: where an order is up to.

The one unit that is reached both ways. It *subscribes* to courierAssigned,
because a courier having been assigned is a statement nobody waits on, and it
is *called* by the storefront, because a customer asking where their food is
wants an answer now.

Both arrive at the same Lambda and the generated entrypoint tells them apart by
the shape of the event. Nothing in this file has to know that.
"""
# Injected by cloudcc: runtime clients for this program's declared capabilities.
from _cloudcc_runtime import kv as _cloudcc_kv


import boto3


from .events import courier_assigned

None

timeline = _cloudcc_kv.connect("trackingEvents")


def on_courier_assigned(message: dict) -> dict:
    """A subscriber, not a remote function: nobody is waiting on the result.

    So it is an ordinary ``def``. The compiler only requires ``async def`` of
    functions something calls over the wire, which this is not -- the message
    arrives on its own.
    """
    order_id = message.get("order_id", "unknown")
    timeline.put_item(
        Item={
            "id": order_id,
            "state": "out-for-delivery",
            "courier": message.get("courier", "unassigned"),
        }
    )
    return {"tracked": order_id}


courier_assigned.subscribe(on_courier_assigned)


async def order_status(order_id: str) -> dict:
    """Called by the storefront while a customer is looking at the page."""
    item = timeline.get_item(Key={"id": order_id}).get("Item")
    if item is None:
        return {"order_id": order_id, "state": "preparing", "courier": None}
    return {
        "order_id": order_id,
        "state": item["state"],
        "courier": item.get("courier"),
    }
